### PR TITLE
add deprecation policy for k8s metrics

### PR DESCRIPTION
**Pull request by logicalhan**
Metric stability is now GA, updating deprecation policy accordingly.

### CODE REVIEW

File: content/en/docs/reference/using-api/deprecation-policy.md [hunk 1]
```diff
@@ -428,6 +428,49 @@ transitions a lifecycle stage as follows. Feature gates must function for no les
...
+are determined by the perceived importance of the metric. The rules for
+deprecating and removing a metric are as follows:
+
+**Rule #9: ALPHA metrics have no guarantees and can be removed at any time.**
+
+**Rule #10: STABLE metrics must undergo a deprecation lifecycle prior to removal.**
+
+   * **STABLE metric: 3 releases**
+   * **STABLE (but deprecated): 2 releases**
+   * **STABLE (but deprecated and now hidden by default): 1 release**
```
Comment (lavalamp) [inline content/en/docs/reference/using-api/deprecation-policy.md:451]: This list isn't super clear, are these separate steps? Or do they overlap?

Comment (logicalhan) [reply content/en/docs/reference/using-api/deprecation-policy.md:451]: ah, they overlap. I'll rephrase.

File: content/en/docs/reference/using-api/deprecation-policy.md [hunk 2]
```diff
@@ -428,6 +428,49 @@ transitions a lifecycle stage as follows. Feature gates must function for no les
...
+
+Deprecated metrics have the same stability guarantees of their counterparts. If a stable
+metric is deprecated, then a deprecated stable metric is guaranteed to not change. When
```
Comment (lavalamp) [inline content/en/docs/reference/using-api/deprecation-policy.md:454]: this sentence is super confusing but I don't know how to help it. Maybe don't make "alpha and deprecated" to be a possible thing. Alpha metrics just get removed, not deprecated. So all "deprecated" metrics must have once been stable, and therefore can't be changed, just like other stable metrics.

File: content/en/docs/reference/using-api/deprecation-policy.md [hunk 3]
```diff
@@ -428,6 +428,49 @@ transitions a lifecycle stage as follows. Feature gates must function for no les
...
+deprecating a stable metric, a future Kubernetes release is specified as the point from
+which the metric will be considered deprecated.
+
+Deprecated metrics will have their description text prefixed with a deprecation notice
+string '(Deprecated from x.y)' and a warning log will be emitted during metric
```
Comment (lavalamp) [inline content/en/docs/reference/using-api/deprecation-policy.md:459]: What is the point to this warning log? consumers of the metrics aren't going to be looking at the logs at that time.

File: content/en/docs/reference/using-api/deprecation-policy.md (hunk 2 above)
Comment (logicalhan) [reply content/en/docs/reference/using-api/deprecation-policy.md:454]: sgtm

**Comment by lavalamp** [comment]
2 releases is pretty short for this. 4 would align more with api deprecation.

I think you should get someone other than me from sig arch to sign off too. Maybe @liggitt.

File: content/en/docs/reference/using-api/deprecation-policy.md (hunk 3 above)
Comment (logicalhan) [reply content/en/docs/reference/using-api/deprecation-policy.md:459]: I disagree actually. Cluster admins also tend to also be consumers of metrics, so they may end up looking at the logs.

**Comment by logicalhan** [comment]
> 2 releases is pretty short for this. 4 would align more with api deprecation.
> 
> I think you should get someone other than me from sig arch to sign off too. Maybe @liggitt.

It's actually 3 releases. But yeah we can modify it to be 4.

**Comment by logicalhan** [comment]
I guess it depends where you are counting from though.

File: content/en/docs/reference/using-api/deprecation-policy.md (hunk 3 above)
Comment (lavalamp) [reply content/en/docs/reference/using-api/deprecation-policy.md:459]: Sure, I don't object to logging them, but I can't believe it's going to happen very often that the person in charge of metrics ingestion is going to happen to be reading the logs and learn about the deprecation there for the first time, in advance of their metrics ingestion breaking.

**Comment by lavalamp** [comment]
Yeah I was counting until they disappear by default.

**Comment by sftim** [comment]
Is this a change for v1.21 or for v1.20 @logicalhan ?

(If it's for v1.21, the PR should target _dev-1.21_ - _master_ represents the live website).

**Comment by logicalhan** [comment]
> Is this a change for v1.21 or for v1.20 @logicalhan ?
> 
> (If it's for v1.21, the PR should target _dev-1.21_ - _master_ represents the live website).

Updated :)

**Comment by reylejano** [comment]
/assign
Hi @logicalhan, is this related to kubernetes/enhancements#1209

**Comment by logicalhan** [comment]
> /assign
> Hi @logicalhan, is this related to [kubernetes/enhancements#1209](https://github.com/kubernetes/enhancements/issues/1209)

It is indeed.

File: content/en/docs/reference/using-api/deprecation-policy.md [hunk 4]
```diff
@@ -428,6 +428,49 @@ transitions a lifecycle stage as follows. Feature gates must function for no les
...
+Accordingly, metrics fall under two stability classes (`ALPHA` and `STABLE`);
+this impacts removal of a metric during a Kubernetes release. These classes
+are determined by the perceived importance of the metric. The rules for
+deprecating and removing a metric are as follows:
+
+**Rule #9: ALPHA metrics have no guarantees and can be removed at any time.**
+
+**Rule #10: STABLE metrics must undergo a deprecation lifecycle prior to removal.**
+
+   * **STABLE metric: 3 releases**
```
Comment (liggitt) [inline content/en/docs/reference/using-api/deprecation-policy.md:449]: Suggest a `X releases or X months, whichever is longer` approach (matching GA API lifetimes would make sense to me)

Also, the 12 month guaranteed lifetime for GA APIs looks increasingly insufficient. We have enormous difficulty getting people to move from *beta* APIs to GA APIs in a year, let alone moving from one GA API to another. I think the promise of GA APIs (or "stable" metrics in this case) is effectively *forever*

Comment (logicalhan) [reply content/en/docs/reference/using-api/deprecation-policy.md:449]: We have a built-in forcing function for migration. We break people but allow them to unbreak themselves for a release to migrate.

Comment (liggitt) [reply content/en/docs/reference/using-api/deprecation-policy.md:449]: > We break people but allow them to unbreak themselves for a release to migrate.

That doesn't match my understanding of "stable" or "GA".

File: content/en/docs/reference/using-api/deprecation-policy.md [hunk 5]
```diff
@@ -428,6 +428,49 @@ transitions a lifecycle stage as follows. Feature gates must function for no les
...
+deprecating a stable metric, a future Kubernetes release is specified as the point from
+which the metric will be considered deprecated.
+
+Deprecated metrics will have their description text prefixed with a deprecation notice
+string '(Deprecated from x.y)' and a warning log will be emitted during metric
+registration. Like their stable undeprecated counterparts, deprecated metrics will
+be automatically registered to the metrics endpoint and therefore visible.
+
+On a subsequent release (when the metric's deprecatedVersion is equal to
+current_kubernetes_version - 1)), a deprecated metric will become a hidden metric.
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-policy.md:464]:
```suggestion
_current_kubernetes_version - 1_)), a deprecated metric will become a hidden metric.  
```
Is that better?

**Comment by reylejano** [comment]
@kubernetes/sig-instrumentation-approvers , please provide a technical review for this PR by March 31 to get this into the release. Thank you!

**Comment by PI-Victor** [comment]
Hi @brancz @x13n @DirectXMan12 @lavalamp @dashpole  @ehashman @mml  @sttts @bsalamat @andrewsykim you are listed as reviewers for [KEP-1209: Metrics Stability Framework](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1209-metrics-stability) i would like to ask you to please provide a tech review of this PR by March 31st in order for it to make it into the release.
Thank you!

**Comment by reylejano** [comment]
Hi @logicalhan, please address [liggit's comment](https://github.com/kubernetes/website/pull/27202#discussion_r601678298) and [sftim's comment](https://github.com/kubernetes/website/pull/27202#discussion_r602557649)

**Comment by logicalhan** [comment]
> Hi @logicalhan, please address [liggit's comment](https://github.com/kubernetes/website/pull/27202#discussion_r601678298) and [sftim's comment](https://github.com/kubernetes/website/pull/27202#discussion_r602557649)

Amended to address comments.

File: content/en/docs/reference/using-api/deprecation-policy.md [hunk 6]
```diff
@@ -428,6 +428,46 @@ transitions a lifecycle stage as follows. Feature gates must function for no les
...
+
+   * **STABLE: 3 releases or 9 months (whichever is longer)**
+   * **ALPHA: 0 releases**
+
+Deprecated metrics will have their description text prefixed with a deprecation notice
+string '(Deprecated from x.y)' and a warning log will be emitted during metric
+registration. Like their stable undeprecated counterparts, deprecated metrics will
+be automatically registered to the metrics endpoint and therefore visible.
+
+On a subsequent release (when the metric's deprecatedVersion is equal to
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-policy.md:460]: nit:
```suggestion
On a subsequent release (when the metric's `deprecatedVersion` is equal to
```

File: content/en/docs/reference/using-api/deprecation-policy.md [hunk 7]
```diff
@@ -428,6 +428,46 @@ transitions a lifecycle stage as follows. Feature gates must function for no les
...
+_current_kubernetes_version - 3_)), a deprecated metric will become a _hidden_ metric.
+**_Unlike_** their deprecated counterparts, hidden metrics will _no longer_ be
+automatically registered to the metrics endpoint (hence hidden). However, they
+can be explicitly enabled through a command line flag on the binary
+(i.e. `--show-hidden-metrics-for-version=`). This provides cluster admins an
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-policy.md:465]: nit: we avoid Latin abbreviations
```suggestion
( `--show-hidden-metrics-for-version=`). This provides cluster admins an
```

**Review by sftim — COMMENTED** [review]
Just for the record

**Comment by sftim** [comment]
Thanks @logicalhan

**Comment by reylejano** [comment]
PR kubernetes/website#27358 is a follow-up PR to address https://github.com/kubernetes/website/pull/27202#pullrequestreview-625177312

**Pull request merged**